README.md
# Shell-in-c
## Para ejecutar el proyecto :
```bash
gcc -o main main.c
./main
```
# Funcionalidades implementadas:
- basic
- multi-pipe
- spaces
- chain
- if
- help
- ctrl+c
- background
- history
- variables

Para más informacion del funcionamiento ejecute el comando help.

help/history.txt
En este apartado nos encontramos los comando ( history, again ), con los cuales podemos llevar un registro
de los ultimos comandos ejecutados.

El comando history imprime en la consola los ultimos 10 comandos ejecutados, siendo el comando 1 el menos reciente.

Ejemplo:

myshell $ history
1: help
2: exit
3: ls
4: cd ..
5: cd probando
6: mv main.c ..
7: ls
8: sleep 10
9: fg 30192
10: jobs

Implementacion:

Una vez finalizada la ejecucion de un comando, se guarda la linea de entrada en un archivo `history` alojado en /home.
Si el archivo contiene mas de 10 comandos, se elimina el menos reciente y se agrega el nuevo. Un problema de la
implementacion es que el comando history no se ve reflejado en el historial en la propia llamada que se le hace
y esto es debido a que los comandos solo se guradan en el historial una vez hayan sido ejecutados, esto lo hacemos
para evitar guardar comandos que condujeron a errores o entradas incorrectas.

--------------------

El comando again <number> ejecuta nuevamente el comando <number> en la lista del history. Si es llamado sin 
parametros, se ejecuta el ultimo comando del history.

Ejemplos:

myshell $ again 3
myshell $ again

Implementacion:

Una vez identificado que se trata del comando again, la entrada es sustituida por el comando correspondiente en el
historial y ejecutado.


execute.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <fcntl.h>
#include "builtin.c"

int execute(char **args, int fdin, int fdout, int master);

// Launch a program.
int shell_launch(char **args, int fdin, int fdout) {
  dup2(fdin, 0);
  dup2(fdout, 1);
  
  if (execvp(args[0], args) == -1) {
    printf("Command not found\n");
    return -1;
  }
}

// Check < value in args
int red_in(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], "<") == 0) {
      args[i] = NULL;
      int fd = open(args[i+1], O_RDWR | O_CREAT, 0644);
      execute(args, fd, fdout, 0);
      
      return 0;
    }
  }

  return -1;
}

// Check > or >> value in args
int red_out(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], ">") == 0) {
      if (i == 0) {
        FILE *f = fopen(args[1], "w");
        fclose(f);
      }
      else {
        args[i] = NULL;
        int fd = open(args[i+1], O_RDWR | O_CREAT, 0644);
        execute(args, fdin, fd, 0);
      }
      
      return 0;
    }
    else if (strcmp(args[i], ">>") == 0) {
      args[i] = NULL;
      int fd = open(args[i+1], O_RDWR | O_CREAT | O_APPEND, 0644);
      execute(args, fdin, fd, 0);

      return 0;
    }
  }

  return -1;
}

// Check pipes
int pipes(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], "|") == 0) {
      char **a_bef = arr_cpy(args, i, 1);
      char **a_aft = arr_cpy(args, i+1, 0);
      pid_t pid;
      int fd[2];
      int in = dup(0);
      int out = dup(1);
      pipe(fd);

      pid = fork();
      if (pid == 0) {
        close(fd[0]);
        shell_launch(a_bef, fdin, fd[1]);
        close(fd[1]);
      } else {
        wait(NULL);
        close(fd[1]);
        execute(a_aft, fd[0], fdout, 0);
        close(fd[0]);
      }

      dup2(in, 0);
      dup2(out, 1);
      close(in);
      close(out);

      return 0;
    }
  }

  return -1;
}

// Chain commands

// Chain ;
int chain(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], ";") == 0) {
      char **a_bef = arr_cpy(args, i, 1);
      char **a_aft = arr_cpy(args, i+1, 0);
      execute(a_bef, fdin, fdout, 0);
      execute(a_aft, fdin, fdout, 0);
      return 0;
    }
  }

  return -1;
}

// Chain &&
int chain_and(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], "&&") == 0) {
      char **a_bef = arr_cpy(args, i, 1);
      char **a_aft = arr_cpy(args, i+1, 0);
      if (execute(a_bef, fdin, fdout, 0) == 0 && execute(a_aft, fdin, fdout, 0) == 0)
        return 0;
      return 1;
    }
  }

  return -1;
}

// Chain ||
int chain_or(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], "||") == 0) {
      char **a_bef = arr_cpy(args, i, 1);
      char **a_aft = arr_cpy(args, i+1, 0);
      if (execute(a_bef, fdin, fdout, 0) == 0 || execute(a_aft, fdin, fdout, 0) == 0)
        return 0;
      return 1;
    }
  }

  return -1;
}

// Conditional execution
int conditions(char **args, int fdin, int fdout) {
  if (strcmp(args[0], "if") == 0) {
    int i = 1;
    args = arr_cpy(args, i, 0);
    i = 0;
    while (strcmp(args[i], "then") != 0) i++;
    char **a_if = arr_cpy(args, i, 1);
    args = arr_cpy(args, i+1, 0);
    i = 0;
    while (strcmp(args[i], "else") != 0) {
      i++;
      if (args[i] == NULL) {
        char **a_then = arr_cpy(args, i-1, 1);
        if (execute(a_if, fdin, fdout, 0) == 0) execute(a_then, fdin, fdout, 0);

        return 0;
      }
    }
    char **a_then = arr_cpy(args, i, 1);
    args = arr_cpy(args, i+1, 0);
    i = 0;
    while (strcmp(args[i], "end") != 0) i++;
    char **a_else = arr_cpy(args, i, 1);
    if (execute(a_if, fdin, fdout, 0) == 0) execute(a_then, fdin, fdout, 0);
    else execute(a_else, fdin, fdout, 0);

    return 0;
  }

  return -1;
}

// Check background
int background(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], "&") == 0) {
      args[i] = NULL;
      int pid = fork();
      if (pid == 0) {
        setpgid(0, 0);
        execute(args, fdin, fdout, 0);
        exit(0);
      }
      if (pid > 0) {
        setpgid(pid, pid);
        char *path = home_dir("jobs.txt");
        FILE *f = fopen(path, "a");
        char *command = concat_array(args, ' ');
        fprintf(f, "%d %s\n", pid, command);
        fclose(f);
        printf("[%d]\t%d\n", count_lines(path), pid);
        if (args[i+1] != NULL) {
          char **a_aft = arr_cpy(args, i+1, 0);
          execute(a_aft, fdin, fdout, 0);
        }
      }
      
      return 0;
    }
  }

  return -1;
}

// Check set
int set(char **args, int fdin, int fdout) {
  char buffer[1024];
  char *var_dir = home_dir("variables.txt");

  if (strcmp(args[0], "set") == 0) {
    if (args[1] == NULL) {
      FILE *f = fopen(var_dir, "r");
      char line[128];
      while (fgets(line, 128, f) != NULL) {
        if (line[0] == '*') continue;
        printf("%s", line);
      }

      return 0;
    }
    else if (args[2] == NULL) {
      printf("set: not enough arguments\n");
      return 0;
    }
    else {
      int last = 2;
      char **a = arr_cpy(args, 2, 0);
      while (args[last+1] != NULL) last++;
      if (args[2][0] == '`' && args[last][strlen(args[last])-1] == '`') {
        if (last == 2) 
          a[0] = sub_str(a[0], 1, strlen(a[0])-2);
        else {
          a[0] = sub_str(a[0], 1, strlen(a[0])-1);
          a[last-2] = sub_str(a[last-2], 0, strlen(a[last-2])-2);
        }

        char *output_dir = home_dir("output.txt");
        int fd = open(output_dir, O_CREAT | O_WRONLY | O_TRUNC, 0644);
        execute(a, fdin, fd, 0);
        close(fd);
        fd = open(output_dir, O_RDONLY);
        read(fd, buffer, 1024);
        buffer[count_chars(output_dir)] = '*';
        buffer[count_chars(output_dir)+1] = '\0';
        close(fd);
      }
      else {
        char *command = concat_array(a, ' ');
        for (int i = 0; i < strlen(command); i++)
          buffer[i] = command[i];
        buffer[strlen(command)] = '\n';
        buffer[strlen(command)+1] = '*';
        buffer[strlen(command)+2] = '\0';
      }

      char *var = args[1];
      delete_var(var);
      FILE *f = fopen(var_dir, "a");
      fprintf(f, "%s = %s\n", var, buffer);
      fclose(f);

      return 0;
    }
  }

  return -1;
}

int (*operators[]) (char **, int, int) = {
  &set,
  &background,
  &chain,
  &conditions,
  &chain_and,
  &chain_or,
  &pipes,
  &red_in,
  &red_out
};

int lsh_num_operators() {
  return sizeof(operators) / sizeof(char *);
}

// Execute shell built-in or launch program.
int execute(char **args, int fdin, int fdout, int master) {
  if (strcmp(args[0], "again") == 0) {
    char *a[3] = {args[0], args[1], NULL};
    args = shell_again(a);
  }
  if (args[0] == NULL) {
    printf("An empty command was entered, don't be a fool.\n");
    return -1;
  }
  // Search for set, &, ;, &&, ||, conditions, |, <, >, >>, and execute them
  for (int i = 0; i < lsh_num_operators(); i++) {
    int result = (*operators[i])(args, fdin, fdout);
    if (result != -1) 
      return master ? 1 : result;
  }
  for (int i = 0; i < lsh_num_builtins(); i++) {
    if (strcmp(args[0], builtin_str[i]) == 0) {
      int result = (*builtin_func[i])(args);
      return master ? 1 : result;
    }
    if (strcmp(args[0], "exit") == 0) return 0;
  }
  int pid = fork();
  if (pid == 0) {
    if (shell_launch(args, fdin, fdout) == -1)
      exit(-1);
  }
  else {
    int status;
    wait(&status);
    if (WIFEXITED(status)) {
      int exit_status = WEXITSTATUS(status);
      if (exit_status == 255) 
        return -1;
    }
  }
  
  return master ? 1 : 0;
}

lsh_functions.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include "execute.c"

char *lsh_read_line(void) {
  char *line = NULL;
  ssize_t bufsize = 0; // have getline allocate a buffer for us

  if (getline(&line, &bufsize, stdin) == -1){
    if (feof(stdin)) {
      exit(EXIT_SUCCESS);  // We recieved an EOF
    } else  {
      perror("readline");
      exit(EXIT_FAILURE);
    }
  }

  return line;
}
char *decod_line(char *line) {
    char *aux_line = (char *) malloc(3 * strlen(line));

    int len = (int) strlen(line);
    int j = 0;
    for (int i = 0; i < len; i++) {
        if (line[i] == '#')
            break;

        if (line[i] == ' ') {
            if (i != 0) {
                if (line[i - 1] == ' ') continue;
            } else
                continue;
        }

        if (line[i] == ';') {
            if (i != 0) {
                if (line[i - 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != len - 1) {
                if (line[i + 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '<') {
            if (i != 0) {
                if (line[i - 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != len - 1) {
                if (line[i + 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '>') {
            if (i != 0) {
                if (line[i - 1] != ' ' && line[i - 1] != '>') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != sizeof(line) - 1) {
                if (line[i + 1] != ' ' && line[i + 1] != '>') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '|') {
            if (i != 0) {
                if (line[i - 1] != ' ' && line[i - 1] != '|') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != len - 1) {
                if (line[i + 1] != ' ' && line[i + 1] != '|') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '&') {
            if (i != len - 1 && i != 0) {
                if (line[i - 1] != ' ' && line[i + 1] == '&') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != 0 && i != len - 1) {
                if (line[i + 1] != ' ' && line[i - 1] == '&') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        aux_line[j++] = line[i];
    }

    char *new_line = (char *) malloc(j + 1);

    for (int x = 0; x < j; x++) {
        new_line[x] = aux_line[x];
    }
    free(aux_line);

    if (new_line[j - 1] != '\n') new_line[j++] = '\n';
    new_line[j] = 0;

    return (char *) new_line;
}

#define LSH_TOK_BUFSIZE 64
#define LSH_TOK_DELIM " \t\r\n\a"
char **lsh_split_line(char *line) {
  char *decoded_line = decod_line(line);
  int bufsize = LSH_TOK_BUFSIZE, position = 0;
  char **tokens = malloc(bufsize * sizeof(char*));
  char *token;

  if (!tokens) {
    fprintf(stderr, "lsh: allocation error\n");
    exit(EXIT_FAILURE);
  }

  token = strtok(decoded_line, LSH_TOK_DELIM);
  while (token != NULL) {
    tokens[position] = token;
    position++;

    if (position >= bufsize) {
      bufsize += LSH_TOK_BUFSIZE;
      tokens = realloc(tokens, bufsize * sizeof(char*));
      if (!tokens) {
        fprintf(stderr, "lsh: allocation error\n");
        exit(EXIT_FAILURE);
      }
    }

    token = strtok(NULL, LSH_TOK_DELIM);
  }
  tokens[position] = NULL;
  return tokens;
}

void lsh_loop(void) {
  char *line;
  char **args;
  int status;

  do {
    printf("myshell$ ");
    line = lsh_read_line();
    args = lsh_split_line(line);
    status = execute(args, 0, 1);

    free(line);
    free(args);
  } while (status);
}

help_functions.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define MY_SH_TOK_DELIM " \t\r\n\a"
#define ERROR "\033[1;31mmy_sh\033[0m"
#define LSH_TOK_BUFSIZE 64

// arr copy
char **arr_cpy(char **arr, int i, int bool) {
  int bufsize = LSH_TOK_BUFSIZE, position = 0;
  char **a_c = malloc(bufsize * sizeof(char*));

  if (bool) {
    while (position < i) {
      a_c[position] = arr[position];
      position++;
    }
    a_c[i] = NULL;
  }
  else {
    while (arr[i] != NULL) 
      a_c[position++] = arr[i++];
    a_c[position] = NULL;
  }
  
  return a_c;
}

// count lines in a file
int count_lines(char *path) {
    FILE *file = fopen(path, "r");
    int lines = 0;
    int c;
    while ((c = fgetc(file)) != EOF) {
        if (c == '\n') {
            lines++;
        }
    }
    fclose(file);
    return lines;
}

// count chars in a file
int count_chars(char *path) {
    FILE *file = fopen(path, "r");
    int chars = 0;
    int c;
    while ((c = fgetc(file)) != EOF) {
        chars++;
    }
    fclose(file);
    return chars;
}

// concat the elements of an array
char* concat_array(char** array, char separator) {
  int array_size = 0;
  while (array[array_size] != NULL)
    array_size++;
  int total_length = 0;
  for (int i = 0; i < array_size; i++) {
    total_length += strlen(array[i]);
      if (i != array_size - 1) {
        total_length++;
      }
  }

  char* result = (char*) malloc((total_length + 1) * sizeof(char));

  int index = 0;
  for (int i = 0; i < array_size; i++) {
      int len = strlen(array[i]);
      memcpy(result + index, array[i], len);
      index += len;
      if (i != array_size - 1) {
          result[index++] = separator;
      }
  }

  result[total_length] = '\0';

  return result;
}

// return a substring
char *sub_str(char *line, int init, int end) {
  char *new_line = (char *) malloc(end - init + 1);

  int i;
  for (i = 0; i < end - init + 1; i++) {
      new_line[i] = line[i + init];
  }
  new_line[i] = 0;

  return new_line;
}

// return the pid of a process
char *get_pid(char *line) {
  int i;
  for (i = 0; i < strlen(line); i++) {
    if (line[i] == ' ') break;
  }
  return sub_str(line, 0, i-1);
}

// return a direction at home 
char *home_dir(char *file_name) {
  char *home_dir = getenv("HOME");
  char *dir[3] = {home_dir, file_name, NULL};
  return concat_array(dir, '/');
}

// delete a variable
void delete_var(char *var) {
  char *var_dir = home_dir("variables.txt");
  FILE *f = fopen(var_dir, "r");
  char *lines[100];
  char line[128];
  int i = 0;
  while (fgets(line, 128, f) != NULL) {
    int index = 0;
    while (line[index] != '\n' && line[index] != '=') index++;
    if (line[index] == '=') {
      char *v = sub_str(line, 0, index - 2);
      if (strcmp(v, var) == 0) {
        do 
          fgets(line, 128, f);
        while(line[0] != '*');
        continue;
      }
    }
    lines[i] = sub_str(line, 0, strlen(line) - 1);
    i++;
  }
  fclose(f);
  f = fopen(var_dir, "w");
  for (int j = 0; j < i; j++) {
    fprintf(f, "%s", lines[j]);
  }
  fclose(f);
}

loop_functions.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <signal.h>
#include <pwd.h>
#include "execute.c"

#define BLUE "\033[1;34m"
#define RESET "\033[0m"
#define GREEN "\033[1;32m"

int count = 0;

char *read_line(void) {
  char *line = NULL;
  ssize_t bufsize = 0; // have getline allocate a buffer for us

  if (getline(&line, &bufsize, stdin) == -1){
    if (feof(stdin)) {
      exit(EXIT_SUCCESS);  // We recieved an EOF
    } else  {
      perror("readline");
      exit(EXIT_FAILURE);
    }
  }

  return line;
}

char *decod_line(char *line) {
    char *aux_line = (char *) malloc(3 * strlen(line));

    int len = (int) strlen(line);
    int j = 0;
    for (int i = 0; i < len; i++) {
        if (line[i] == '#')
            break;

        if (line[i] == ' ') {
            if (i != 0) {
                if (line[i - 1] == ' ') continue;
            } else
                continue;
        }

        if (line[i] == ';') {
            if (i != 0) {
                if (line[i - 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != len - 1) {
                if (line[i + 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '<') {
            if (i != 0) {
                if (line[i - 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != len - 1) {
                if (line[i + 1] != ' ') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '>') {
            if (i != 0) {
                if (line[i - 1] != ' ' && line[i - 1] != '>') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != sizeof(line) - 1) {
                if (line[i + 1] != ' ' && line[i + 1] != '>') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '|') {
            if (i != 0) {
                if (line[i - 1] != ' ' && line[i - 1] != '|') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != len - 1) {
                if (line[i + 1] != ' ' && line[i + 1] != '|') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        if (line[i] == '&') {
            if (i != len - 1 && i != 0) {
                if (line[i - 1] != ' ' && line[i + 1] == '&') {
                    aux_line[j++] = ' ';
                }
            }
            aux_line[j++] = line[i];
            if (i != 0 && i != len - 1) {
                if (line[i + 1] != ' ' && line[i - 1] == '&') {
                    aux_line[j++] = ' ';
                }
            }

            continue;
        }

        aux_line[j++] = line[i];
    }

    char *new_line = (char *) malloc(j + 1);

    for (int x = 0; x < j; x++) {
        new_line[x] = aux_line[x];
    }
    free(aux_line);

    if (new_line[j - 1] != '\n') new_line[j++] = '\n';
    new_line[j] = 0;

    return (char *) new_line;
}

#define LSH_TOK_BUFSIZE 64
#define LSH_TOK_DELIM " \t\r\n\a"
char **split_line(char *line) {
  char *decoded_line = decod_line(line);
  int bufsize = LSH_TOK_BUFSIZE, position = 0;
  char **tokens = malloc(bufsize * sizeof(char*));
  char *token;

  if (!tokens) {
    fprintf(stderr, "lsh: allocation error\n");
    exit(EXIT_FAILURE);
  }

  token = strtok(decoded_line, LSH_TOK_DELIM);
  while (token != NULL) {
    tokens[position] = token;
    position++;

    if (position >= bufsize) {
      bufsize += LSH_TOK_BUFSIZE;
      tokens = realloc(tokens, bufsize * sizeof(char*));
      if (!tokens) {
        fprintf(stderr, "lsh: allocation error\n");
        exit(EXIT_FAILURE);
      }
    }

    token = strtok(NULL, LSH_TOK_DELIM);
  }
  tokens[position] = NULL;
  return tokens;
}

void manage_ctrl_c(int signal) {
    printf("\n");
    if (count) exit(signal);
    count++;
}

void write_history(char **args) {
    if (strcmp(args[0], "again") == 0) {
        char *a[3] = {args[0], args[1], NULL};
        args = shell_again(a);
    }
    char *history_dir = home_dir("history.txt");
    int c_lines = count_lines(history_dir);
    char *command = concat_array(args, ' ');
    FILE *f = fopen(history_dir, "r");
    char *lines[9];

    if (c_lines == 10) {
        char line[128];
        fgets(line, 128, f);
        for (int i = 0; i < 9; i++) {
            fgets(line, 128, f);
            lines[i] = sub_str(line, 0, strlen(line) - 1);
        } 
        f = fopen(history_dir, "w");
        for (int i = 0; i < 9; i++) 
            fprintf(f, "%s", lines[i]);
        fclose(f);
    } 
    f = fopen(history_dir, "a");
    fprintf(f, "%s\n", command);
    fclose(f);
}

void loop() {
  char *line;
  char **args;
  int status = 1;
  char cwd[128];
  char *prompt = "\033[0;32mmyshell\033[0m";
  struct passwd *pw = getpwuid(getuid());

  signal(SIGINT, manage_ctrl_c);
  while (status) {
    getcwd(cwd, sizeof(cwd));
    printf("%smyshell@%s%s:%s~%s%s$ ", GREEN, pw->pw_name, RESET, BLUE, cwd, RESET);
    line = read_line();
    args = split_line(line);
    status = execute(args, 0, 1, 1);
    count = 0;
    if (status != -1)
        write_history(args);
    free(line);
    free(args);
  }
}

pipes_redirectors.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <fcntl.h>
#include "help.c"

// Check < value in args
int check_red_in(char **args) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], "<") == 0) return i;
  }

  return 0;
}

// Check > or >> value in args
int check_red_out(char **args) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], ">") == 0) return i;
    else if (strcmp(args[i], ">>") == 0) return -i;
  }

  return 0;
}

// Check pipes
int pipes(char **args, int fdin, int fdout) {
  for (int i = 0; args[i] != NULL; i++) {
    if (strcmp(args[i], "|") == 0) {
      char **a_bef = arr_cpy(args, i, 1);
      char **a_aft = arr_cpy(args, i+1, 0);
      pid_t pid;
      int fd[2];
      int in = dup(0);
      int out = dup(1);
      pipe(fd);

      pid = fork();
      if (pid == 0) {
        close(fd[0]);
        shell_launch(a_bef, fdin, fd[1]);
        close(fd[1]);
      } else {
        wait(NULL);
        close(fd[1]);
        lsh_execute(a_aft, fd[0], fdout);
        close(fd[0]);
      }

      dup2(in, 0);
      dup2(out, 1);
      close(in);
      close(out);

      return 0;
    }
  }

  return 1;
}

builtin.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <dirent.h>
#include "help_functions.c"

// Function Declarations for builtin shell commands:
int shell_cd(char **args);
int shell_help(char **args);
int shell_true(char **args);
int shell_false(char **args);
int shell_jobs(char **args);
int shell_fg(char ** args);
int shell_history(char **args);
int shell_unset(char **args);
int shell_get(char **args);

// List of builtin commands, followed by their corresponding functions.
char *builtin_str[] = {
  "cd",
  "help",
  "true",
  "false",
  "jobs",
  "fg",
  "history",
  "unset",
  "get"
};

int (*builtin_func[]) (char **) = {
  &shell_cd,
  &shell_help,
  &shell_true, 
  &shell_false,
  &shell_jobs,
  &shell_fg,
  &shell_history,
  &shell_unset,
  &shell_get
};

int lsh_num_builtins() {
  return sizeof(builtin_str) / sizeof(char *);
}

// Builtin function implementations.

// Change directory bultin command.
int shell_cd(char **args) {
  if (args[1] == NULL) {
    char *home_dir = getenv("HOME");
    int ret = chdir(home_dir);
  } else {
    if (chdir(args[1]) != 0) {
      perror("lsh");
    }
  }
  return 1;
}

// Help builtin command.
int shell_help(char **args) {
  char *buf = malloc(1000);
  ssize_t bytes_read;
  if (args[1] == NULL) {
    int fd = open("help/general.txt", O_RDONLY);
    ssize_t bytes_read;
    while ((bytes_read = read(fd, buf, 1000)) > 0) {
      write(1, buf, bytes_read);
    }
    free(buf);
    close(fd);
  }
  else {
    char *cmd_name = sub_str(args[1], 0, strlen(args[1]) - 1);
    char help[100] = "help/";
    char *new_str = strcat(help, strcat(cmd_name, ".txt"));
    int fd = open(new_str, O_RDONLY);
    if (fd < 0) {
      int size;
      for (int i = 0; i < 100; i++) {
        if (cmd_name[i] == '.') {
          size = i;
          break;
        }
      }
      char *new_args[3] = {sub_str(cmd_name, 0, size-1), "--help", NULL};
      int pid = fork();
      if (pid == 0) execvp(new_args[0], new_args);
      else wait(NULL);

      return 1;
    }
    while ((bytes_read = read(fd, buf, 1000)) > 0) {
      write(1, buf, bytes_read);
    }
    free(buf);
    close(fd);

    return 1;
  }

  return 1;
}

// True builtin command.
int shell_true(char **args) {
  return 0;
}

// False builtin command.
int shell_false(char **args) {
  return 1;
}

// Jobs builtin command.
int shell_jobs(char **args) {
  char *jobs_dir = home_dir("jobs.txt");
  FILE *fr = fopen(jobs_dir, "r");
  char line[100];
  int line_count = 1;
  int bufsize = 64;
  char *running_jobs[100];;
  int index = 0;

  while (fgets(line, 100, fr) != NULL && strlen(line) > 0) {
    char *pid = get_pid(line);
    char *cmd = sub_str(line, strlen(pid) + 1, strlen(line) - 1);
    int is_alive = waitpid(atoi(pid), NULL, WNOHANG);

    if (is_alive == 0) {
      printf("[%d] +Running\t%s", line_count, cmd);
      running_jobs[index] = sub_str(line, 0, strlen(line) - 1);
      index++;
    }
    else {
      printf("[%d] -Done\t%s", line_count, cmd);
    }
    line_count++;
  }
  running_jobs[index] = NULL;
  fclose(fr);
  FILE *fw = fopen(jobs_dir, "w");
  for (int i = 0; running_jobs[i] != NULL; i++) 
    fprintf(fw, "%s", running_jobs[i]);
  fclose(fw);

  return 1;
}

// Foreground builtin command.
int shell_fg(char **args) {
  FILE *f = fopen(home_dir("jobs.txt"), "r");
  char line[100];
  int pid, w;

  if (args[1] == NULL) {
    char *n_line;
    while (fgets(line, 100, f) != NULL && strlen(line) > 0) 
      n_line = sub_str(line, 0, strlen(line) - 1);
    pid = atoi(get_pid(n_line));    
  }
  else {
    int bool = 0;
    pid = atoi(args[1]);
    while (fgets(line, 100, f) != NULL && strlen(line) > 0) {
      if (pid == atoi(get_pid(line)))
        bool = 1;
      if (bool != 1) {
        printf("The process is not in the background\n");
        return 1;
      }
    }
  }

  if (waitpid(pid, NULL, WNOHANG) == 0) {
      do {
        w = waitpid(pid, NULL, WNOHANG);
      } while (w == 0);
  }
  else printf("Process finished in background\n");
  
  fclose(f);

  return 1;
}

// History builtin command
int shell_history(char **args) {
  char *history_dir = home_dir("history.txt");
  FILE *fr = fopen(history_dir, "r");
  char line[100];
  int line_count = 1;

  while (fgets(line, 100, fr) != NULL && strlen(line) > 0) {
    printf("%d: %s", line_count, line);
    line_count++;
  }
  fclose(fr);

  return 1;
}

// Again builtin command
char **shell_again(char **args) {
  char *history_dir = home_dir("history.txt");
  FILE *fr = fopen(history_dir, "r");
  char line[100];
  int line_count = 1;
  char **new_args = malloc(100 * sizeof(char *));
  int index = 0;

  while (fgets(line, 100, fr) != NULL && strlen(line) > 0) {
    if (atoi(args[1]) == line_count) {
      char *cmd = sub_str(line, 0, strlen(line) - 2);
      char *token = strtok(cmd, " ");
      while (token != NULL) {
        new_args[index] = token;
        token = strtok(NULL, " ");
        index++;
      }
      new_args[index] = NULL;
      fclose(fr);
      return new_args;
    }
    line_count++;
  }
  fclose(fr);

  return NULL;
}

// Unset bultin command
int shell_unset(char **args) {
  if (args[1] == NULL) {
    printf("unset: not enough arguments\n");
    return 1;
  }
  else {
    char *var = args[1];
    delete_var(var);
  }

  return 1;
}

// Get bultin command
int shell_get(char **args) {
  if (args[1] == NULL) 
    return 1;
  else {
    char *var = args[1];
    char *var_dir = home_dir("variables.txt");
    FILE *f = fopen(var_dir, "r");
    char line[128];

    while (fgets(line, 128, f) != NULL) {
      int index = 0;
      while (line[index] != '\n' && line[index] != '=') index++;
      if (line[index] == '=') {
        char *v = sub_str(line, 0, index - 2);
        if (strcmp(v, var) == 0) {
          char *output_dir = home_dir("output.txt");
          FILE *fo = fopen(output_dir, "w");
          char *substr = sub_str(line, index + 2, strlen(line) - 1);
          printf("%s", substr);
          fprintf(fo, "%s", substr);
          while(1) {
            fgets(line, 128, f);
            if (line[0] == '*') break;
            printf("%s", line);
            fprintf(fo, "%s", line);
          }
          fclose(fo);

          return 1;
        }
      }
    }
  }

  return 1;
}